scene/2d/physics_body_2d.cpp

```diff
@@ -1111,72 +1126,149 @@ void CharacterBody2D::move_and_slide() {
 
 			if (collided) {
 				found_collision = true;
-
+				on_moving_surface = false;
 				motion_results.push_back(result);
 				_set_collision_direction(result);
 
-				if (on_floor && stop_on_slope) {
-					if ((body_velocity_normal + up_direction).length() < 0.01) {
+				if (on_floor && stop_on_slope && ((motion.normalized() + up_direction).length() < 0.01)) {
+					Transform2D gt = get_global_transform();
+					if (result.motion.length() > margin) {
+						gt.elements[2] -= result.motion.slide(up_direction);
+					} else {
+						gt.elements[2] -= result.motion;
+					}
+					set_global_transform(gt);
+					linear_velocity = Vector2();
+					return;
+				}
+
+				if (result.remainder.is_equal_approx(Vector2())) {
+					motion = Vector2();
+					break;
+				}
+
+				// move on floor only checks
```

Comment:
Comments should start with uppercase and end with trailing `.` (I know some existing ones don't follow this rule, but better not to add more).